python/diff_frames.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python
+import os
+import sys
+import argh
+import asyncio
+import subprocess
+from collections import namedtuple
+
+import pysndfile
+
+from terminal import LinePrintingTerminal, Keyboard
+
+bdiff_frames_path = os.path.join(
+    os.path.dirname(__file__), os.pardir, 'diff_frames')
+
+Hunk = namedtuple('Hunk', ('start_a', 'end_a', 'start_b', 'end_b'))
+
+
+def duration(sndfile):
+    return sndfile.frames() / sndfile.samplerate()
+
+
+def fmt_seconds(t):
+    hours, remainder = divmod(t, 3600)
+    minutes, seconds = divmod(remainder, 60)
+    if not hours:
+        if not minutes:
+            return '{:.2f}'.format(seconds)
+        return '{:.0f}:{:.1f}'.format(minutes, seconds)
+    return '{:.0f}:{:.0f}:{.0f}'.format(hours, minutes, seconds)
+
+
+class DiffApp:
+    def __init__(self, filename_a, filename_b, loop=None):
+        self.filename_a = filename_a
+        self.filename_b = filename_b
+        self._psf_a = None
+        self._psf_b = None
+        self._diff = None
+        self._len = None
+
+        self._loop = loop or asyncio.get_event_loop()
+        self._terminal = LinePrintingTerminal()
+        self._common_fmt = self._terminal.yellow
+        self._insertion_fmt = self._terminal.green
+        self._loop.add_reader(self._terminal.infile, self._handle_input)
+        self._keyboard = Keyboard()
+        self.bindings = {}
+
+    @asyncio.coroutine
+    def __call__(self):
+        diff = yield from self._do_diff()
+        self._diff, a_offset, b_offset = self._process_diff(diff)
+        self._psf_a = pysndfile.PySndfile(self.filename_a)
```

Comment:
I am wondering whether the file opens (and other things we have to do) would be better in __init__ since then they would be run in the synchronous (and therefore easier for error handling) portion of the code?